node/src/transition_frontier/candidate/transition_frontier_candidate_state.rs

```diff
@@ -89,113 +71,189 @@ impl TransitionFrontierCandidateStatus {
-            let prev_hash = block.status.compared_with()?;
-            let prev = self.blocks.get(prev_hash)?;
-            Some(BlockRef {
-                hash: prev_hash,
-                header: &prev.block.header,
-                body: &prev.block.body.staged_ledger_diff,
-                status: &prev.status,
-            })
+    pub(super) fn invalidate(&mut self, hash: &StateHash) {
+        self.ordered.retain(|s| {
+            if s.block.hash() == hash {
+                self.invalid.insert(hash.clone(), s.block.global_slot());
+                false
+            } else {
+                true
+            }
+        });
+    }
+
+    pub(super) fn set_chain_proof(
+        &mut self,
+        hash: &StateHash,
+        chain_proof: (Vec<StateHash>, ArcBlockWithHash),
+    ) -> bool {
+        self.update(hash, move |mut s| {
+            s.chain_proof = Some(chain_proof);
+            s
         })
     }
 
-    pub fn is_candidate_decided_to_use_as_tip(&self, hash: &StateHash) -> bool {
-        let Some(candidate) = self.blocks.get(hash) else {
-            return false;
+    pub(super) fn prune(&mut self) {
+        let mut has_reached_best_candidate = false;
+        let Some(best_candidate_hash) = self.best_verified().map(|s| s.block.hash().clone()) else {
+            return;
         };
-        match &candidate.status {
-            TransitionFrontierCandidateStatus::Received { .. } => false,
-            TransitionFrontierCandidateStatus::Prevalidated => false,
-            TransitionFrontierCandidateStatus::SnarkVerifyPending { .. } => false,
-            TransitionFrontierCandidateStatus::SnarkVerifySuccess { .. } => false,
-            TransitionFrontierCandidateStatus::ForkRangeDetected { .. } => false,
-            TransitionFrontierCandidateStatus::ShortRangeForkResolve {
-                compared_with,
-                decision,
-                ..
-            } => decision.use_as_best_tip() && &self.best_tip == compared_with,
-            TransitionFrontierCandidateStatus::LongRangeForkResolve {
-                compared_with,
-                decision,
-                ..
-            } => decision.use_as_best_tip() && self.best_tip.as_ref() == Some(compared_with),
-        }
+
+        // prune all blocks that are worse(consensus-wise) than the best
+        // verified candidate.
+        self.ordered.retain(|s| {
+            if s.block.hash() == &best_candidate_hash {
+                // prune all invalid block hashes which are for older
+                // slots than the current best candidate.
+                let best_candidate_slot = s.block.global_slot();
+                self.invalid.retain(|_, slot| *slot >= best_candidate_slot);
+
+                has_reached_best_candidate = true;
+            }
+
+            has_reached_best_candidate
+        });
+    }
+
+    pub(super) fn best(&self) -> Option<&TransitionFrontierCandidateState> {
+        self.ordered.last()
```

Comment:
Just a random comment: It makes perfect sense for the better blocks to be `>` the inferior ones, but for some reason I hate that the best one ends up being last in the collection

Reply:
It's consistent with how blocks are stored in transition frontier as well (best_tip is last), so it should be fine.